src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,1966 @@
+(*
+QCheck: Random testing for OCaml
+copyright (c) 2013-2017, Guillaume Bury, Simon Cruanes, Vincent Hugot, Jan Midtgaard
+all rights reserved.
+*)
+
+(** {1 Quickcheck inspired property-based testing} *)
+
+let poly_compare=compare
+open Printf
+
+module RS = Random.State
+
+let (|>) x f = f x
+
+let rec foldn ~f ~init:acc i =
+  if i = 0 then acc else foldn ~f ~init:(f acc i) (i-1)
+
+let _is_some = function Some _ -> true | None -> false
+
+let _opt_map_2 ~f a b = match a, b with
+  | Some x, Some y -> Some (f x y)
+  | _ -> None
+
+let _opt_map_3 ~f a b c = match a, b, c with
+  | Some x, Some y, Some z -> Some (f x y z)
+  | _ -> None
+
+let _opt_map_4 ~f a b c d = match a, b, c, d with
+  | Some x, Some y, Some z, Some w -> Some (f x y z w)
+  | _ -> None
```

Comment:
I suggest to change the implementation for: 

```ocaml
let _opt_bind f = function 
  | None -> None 
  | Some x -> f x
```
In order to facilitate the implementation of multimap/lift : 

```ocaml
let _opt_lift f a = _opt_bind (fun x -> f (Some x)) a
let _opt_lift2 f ma mb = _opt_bind (fun a -> _opt_bind (fun b -> Some (f a b)) mb) ma)
let _opt_lift3 f ma mb mc = bind (fun a -> bind (fun b -> bind (fun c -> Some (f a b c)) mc) mb) ma
```
But I guess that you can define an internal module and relay on operators for more clarity? (And gives a proper indentation).

Reply:
Oh it was the default implementation, so we can resolve this comment. Sorry for the noise.